torch/csrc/jit/fuser/cpu/fused_kernel.cpp

```diff
@@ -27,6 +53,96 @@ static bool programExists(const std::string& program) {
   return (system(cmd.c_str()) == 0);
 }
 
+#ifdef _MSC_VER
+std::string exec(const std::string& cmd) {
+  std::array<char, 128> buffer;
+  std::string result;
+  std::unique_ptr<FILE, decltype(&_pclose)> pipe(
+      _popen(cmd.c_str(), "r"), _pclose);
+  if (!pipe) {
+    throw std::runtime_error("popen() failed!");
+  }
+  while (fgets(buffer.data(), static_cast<int>(buffer.size()), pipe.get()) != nullptr) {
+    result += buffer.data();
+  }
+  return result;
+}
+
+inline std::string& rtrim(std::string& s, const char* t = " \t\n\r\f\v") {
+  s.erase(s.find_last_not_of(t) + 1);
+  return s;
+}
+
+void activate() {
+  char* root = nullptr;
+  std::string cmd;
+  std::string path;
+  std::string vcruntime_plat;
+  std::string envvars;
+
+  // Checking whether the environment is already activated
+  if (getenv("VSCMD_ARG_TGT_ARCH")) {
+    return;
+  }
+
+  // Getting `ProgramFiles` through environment variable queries
+  root = getenv("ProgramFiles(x86)");
+  if (!root) {
+    root = getenv("ProgramFiles");
+  }
+  if (!root) {
+    return;
+  }
+
+  // Getting VS 2017 installation path through `vswhere`
+  cmd = "\"" + std::string(root) +
+      "\\Microsoft Visual Studio\\Installer\\vswhere.exe\""
+      " -latest -prerelease -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -property installationPath";
+  std::cout << cmd << std::endl;
+  try {
+    path = exec(cmd);
+    rtrim(path);
```

Comment:
Instead of the try-catch here, it would be nicer if `exec` returned an `optional<path>` and then we just test if it's OK and return if it's not.

Reply:
Could you please tell me how to use `optional<path>`? Maybe some examples are welcomed.

Reply:
Something like:

```
c10::optional<std::string> exec(const std::string& cmd) {
  std::array<char, 128> buffer;
  std::string result;
  std::unique_ptr<FILE, decltype(&_pclose)> pipe(
      _popen(cmd.c_str(), "r"), _pclose);
  if (!pipe) {
    return c10::nullopt;
  }
  while (fgets(buffer.data(), static_cast<int>(buffer.size()), pipe.get()) != nullptr) {
    result += buffer.data();
  }
  return result;
}
```

The result can be turned into a string with `*mb_string` and tested as a boolean to see if it is occupied.